pkg/provenance/provenance.go

```diff
@@ -164,105 +170,335 @@ func (o ObjectLineage) SpecHistory() string {
-		}
+	sort.Ints(s)
+	//get all versions, sort by version, make string array of them
+	specs := make([]Spec, 0)
+	for _, version := range s {
+		specs = append(specs, o[version]) //cast Spec to String
+	}
+	specStrings := make([]string, 0)
+	for _, spec := range specs {
+		specStrings = append(specStrings, spec.String())
 	}
-	return strconv.Itoa(-1)
+	return strings.Join(specStrings, "\n")
 }
 
-//TODO: add optional parameters to spechistory route in apiserver.go, and call this method.
-// Right now it is actually unused
-func (o ObjectLineage) SpecHistoryInterval(vNumStart, vNumEnd int) []string {
-	//order keys so we append in order later, reference: https://blog.golang.org/go-maps-in-action#TOC_7.
-	s := make([]Spec, 0)
-	for _, spec := range o {
-		s = append(s, spec)
-	}
-	sort.Slice(s, func(i, j int) bool {
-		return s[i].Version < s[i].Version
-	})
-	specs := make([]string, 0)
-	for _, spec := range s {
-		if spec.Version >= vNumStart && spec.Version <= vNumEnd {
-			specs = append(specs, spec.String())
+func (o ObjectLineage) Bisect(argMap map[string]string) string {
+	s := make([]int, 0)
+	for _, value := range o {
+		s = append(s, value.Version)
+	}
+	sort.Ints(s)
+	//get all versions, sort by version, and then build the spec array in order.
+	//if the query conditions are true more than once, want to find the earliest one.
+	specs := make([]Spec, 0)
+	for _, version := range s {
+		specs = append(specs, o[version])
+	}
+	allQueryPairs := make([][]string, 0)
+	//this section is storing the mappings from the query, into the allQueryPairs object
+	//these are the queries, each element must be satisfied somewhere in the spec for the bisect to succeed
+	for key, value := range argMap {
+		attributeValueSlice := make([]string, 2)
+		if strings.Contains(key, "field") {
+			//find associated value in argMap
+			fieldNum, err := strconv.Atoi(key[5:])
+			if err != nil {
+				return fmt.Sprintf("Failure, could not convert %s. Invalid Query parameters.", key)
+			}
+			//find assocaited value1
+			valueOfKey, ok := argMap["value"+strconv.Itoa(fieldNum)]
+			if !ok {
+				return fmt.Sprintf("Could not find an associated value for field: %s", key)
+			}
+			attributeValueSlice[0] = value
+			attributeValueSlice[1] = valueOfKey
+			allQueryPairs = append(allQueryPairs, attributeValueSlice)
+		}
+	}
+	// fmt.Printf("attributeValuePairs%s\n", allQueryPairs)
+
+	//this section is to build the attributeRelationships from the Query
+	//will use this to ensure that fields belonging to the same top-level
+	//attribute, will be treated as a joint query. So you can't just
+	//ask if username ever is daniel and password is ever 223843, because
+	//it could find that in different parts of the spec. they both must be satisfied in the same map object
+	andGate := make([]bool, len(allQueryPairs))
+	mapRelationships := make(map[string][][]string, 0) // a map from top level attribute to array of pairs (represented as 2 len array)
+	for _, spec := range specs {
```

Comment:
It will be good idea to pull this part out (lines 246 - 271) into its own function. That way we can write unit tests for it.